tools/clang/lib/SPIRV/SPIRVEmitter.cpp

```diff
@@ -2516,6 +2518,140 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
+  // same component type. The only allowed component types are int and float.
+
+  // mul(scalar, scalar)
+  if (TypeTranslator::isScalarType(arg0Type) &&
+      TypeTranslator::isScalarType(arg1Type)) {
+    return theBuilder.createBinaryOp(translateOp(BO_Mul, arg0Type),
+                                     returnTypeId, doExpr(arg0), doExpr(arg1));
+  }
+
+  // mul(scalar, vector)
+  {
+    uint32_t elemCount = 0;
+    if (TypeTranslator::isScalarType(arg0Type) &&
+        TypeTranslator::isVectorType(arg1Type, nullptr, &elemCount)) {
+
+      // We can use OpVectorTimesScalar if arguments are floats.
+      if (arg0Type->isFloatingType())
+        return theBuilder.createBinaryOp(spv::Op::OpVectorTimesScalar,
+                                         returnTypeId, doExpr(arg1),
+                                         doExpr(arg0));
+
+      // Use OpIMul for integers
+      return theBuilder.createBinaryOp(spv::Op::OpIMul, returnTypeId,
+                                       createVectorSplat(arg0, elemCount),
+                                       doExpr(arg1));
+    }
+  }
+
+  // mul(vector, scalar)
+  {
+    uint32_t elemCount = 0;
+    if (TypeTranslator::isVectorType(arg0Type, nullptr, &elemCount) &&
+        TypeTranslator::isScalarType(arg1Type)) {
+
+      // We can use OpVectorTimesScalar if arguments are floats.
+      if (arg1Type->isFloatingType())
+        return theBuilder.createBinaryOp(spv::Op::OpVectorTimesScalar,
+                                         returnTypeId, doExpr(arg0),
+                                         doExpr(arg1));
+
+      // Use OpIMul for integers
+      return theBuilder.createBinaryOp(spv::Op::OpIMul, returnTypeId,
+                                       createVectorSplat(arg1, elemCount),
+                                       doExpr(arg0));
+    }
+  }
+
+  // mul(scalar, matrix)
+  if (TypeTranslator::isScalarType(arg0Type) &&
+      TypeTranslator::isMxNMatrix(arg1Type)) {
+    // We currently only support float matrices. So we can use
+    // OpMatrixTimesScalar
+    if (arg0Type->isFloatingType())
+      return theBuilder.createBinaryOp(spv::Op::OpMatrixTimesScalar,
+                                       returnTypeId, doExpr(arg1),
+                                       doExpr(arg0));
+  }
+
+  // mul(matrix, scalar)
+  if (TypeTranslator::isScalarType(arg1Type) &&
+      TypeTranslator::isMxNMatrix(arg0Type)) {
+    // We currently only support float matrices. So we can use
+    // OpMatrixTimesScalar
+    if (arg1Type->isFloatingType())
+      return theBuilder.createBinaryOp(spv::Op::OpMatrixTimesScalar,
+                                       returnTypeId, doExpr(arg0),
+                                       doExpr(arg1));
+  }
+
+  // mul(vector, vector)
+  if (TypeTranslator::isVectorType(arg0Type) &&
+      TypeTranslator::isVectorType(arg1Type))
+    return processIntrinsicDot(callExpr);
+
+  // mul(vector, matrix)
```

Comment:
I think for mul(vector, matrix), mul(matrix, vector), and mul(matrix, matrix), we should swap the arguments?

https://github.com/google/DirectXShaderCompiler/blob/spirv/docs/SPIR-V.rst#vectors-and-matricesspi
```
mul(vector, matrix) -> OpMatrixTimesVector(arg1, arg0)
mul(matrix, vector) -> OpVectorTimesMatrix(arg1, arg0)
mul(matrix, matrix) -> OpMatrixTimesMatrix(arg1, arg0)
```